### PR TITLE
Make FasterRCNN_mobilenet exportable via torch.export with dynamic shapes

### PR DESCRIPTION
**Pull request by mergennachin**
Partially fixes https://github.com/pytorch/pytorch/issues/146152

Fix data-dependent branches, symbolic shape issues, and unbacked
symbol handling so that fasterrcnn_mobilenet_v3_large_fpn can be
exported with dynamic H/W dimensions using torch.export.export()
in both strict=False and strict=True modes.

All changes preserve eager and JIT behavior exactly by gating
export-specific code behind torch.jit.is_scripting() and
torch.compiler.is_compiling() guards.

Source changes:
- boxes.py: batched_nms dispatch and _batched_nms_coordinate_trick
  empty check made export-safe via guard_or_false
- transform.py: batch_images uses symbolic-safe integer arithmetic
  instead of float ceil
- _utils.py: BoxCoder.decode uses guard_or_true and explicit reshape
  dims to avoid data-dependent guards and -1 ambiguity
- poolers.py: len(rois) -> rois.shape[0] to preserve symbolic ints
- _meta_registrations.py: NMS fake kernel uses new_dynamic_size(min=0)
  to allow empty NMS results at runtime

Tests: 32 new export tests covering NMS, batched_nms,
MultiScaleRoIAlign, BoxCoder.decode, batch_images, and full model
export with dynamic shapes, edge cases, and both strict modes.

### CODE REVIEW

File: torchvision/ops/boxes.py [hunk 1]
```diff
@@ -77,10 +78,17 @@ def batched_nms(
...
-    else:
+    if torch.jit.is_scripting():
+        # _is_tracing() is always False during scripting, so omitted here
+        if boxes.numel() > (4000 if boxes.device.type == "cpu" else 100_000):
+            return _batched_nms_vanilla(boxes, scores, idxs, iou_threshold)
         return _batched_nms_coordinate_trick(boxes, scores, idxs, iou_threshold)
+    # In eager mode, use the original performance-based dispatch.
+    # In export (is_compiling), always use coordinate_trick to avoid
+    # data-dependent branching on boxes.numel().
+    if not torch.compiler.is_compiling() and boxes.numel() > (4000 if boxes.device.type == "cpu" else 100_000) and not torchvision._is_tracing():
```
Comment (laithsakka) [inline torchvision/ops/boxes.py]: is _batched_nms_vanilla not compile friendly or we just avoiding the branches here?
it its just to avoid DDE on the branches i would rather that we do 

```
x = 4000 if boxes.device.type == "cpu" else 100_000
if  not torch.compiler.is_compiling()  and not torchvision._is_tracing() and (guard_or_false(boxes.numel() > x):
..
```

Comment (mergennachin) [reply torchvision/ops/boxes.py]: _batched_nms_vanilla is not compile-friendly. It has for class_id in torch.unique(idxs) which is a data-dependent loop. So we're avoiding both the DDE on the branch condition AND the untraceable vanilla implementation.

File: torchvision/models/detection/_utils.py [hunk 2]
```diff
@@ -173,11 +174,23 @@ def decode(self, rel_codes: Tensor, boxes: list[Tensor]) -> Tensor:
...
+        # guard on unbacked box_sum) and explicit dims (avoids reshape -1
+        # ambiguity when box_sum is 0).
+        if torch.jit.is_scripting():
+            if box_sum > 0:
+                rel_codes = rel_codes.reshape(box_sum, -1)
+            pred_boxes = self.decode_single(rel_codes, concat_boxes)
+            if box_sum > 0:
+                pred_boxes = pred_boxes.reshape(box_sum, -1, 4)
+        else:
+            if guard_or_true(box_sum > 0):
```
Comment (JacobSzwejbka) [inline torchvision/models/detection/_utils.py]: You could have stronger asserts for this function right? Assert that the list of boxes is always greater than 0 (why would someone call decode on empty list?) and that every image in the list has a box. 

In general why export this one over decode_single though?

File: torchvision/ops/boxes.py [hunk 3]
```diff
@@ -94,9 +106,21 @@ def _batched_nms_coordinate_trick(
...
+            return torch.empty((0,), dtype=torch.int64, device=boxes.device)
+        max_coordinate = boxes.max()
+    else:
+        if guard_or_false(boxes.numel() == 0):
+            return torch.empty((0,), dtype=torch.int64, device=boxes.device)
+        if torch.compiler.is_compiling():
+            # Concat a zero sentinel so .max() is safe when boxes is empty
+            # (export traces the non-empty path, but at runtime boxes can be empty)
+            max_coordinate = torch.cat(
+                [boxes.reshape(-1), torch.zeros(1, dtype=boxes.dtype, device=boxes.device)]
```
Comment (JacobSzwejbka) [inline torchvision/ops/boxes.py:121]: This doesnt seem equivalent to me. Does the downstream decoder handle [0] tensor the same as [] ?

File: torchvision/models/detection/_utils.py [hunk 4]
```diff
@@ -173,11 +174,23 @@ def decode(self, rel_codes: Tensor, boxes: list[Tensor]) -> Tensor:
         box_sum = 0
         for val in boxes_per_image:
```
Comment (JacobSzwejbka) [inline torchvision/models/detection/_utils.py:174]: IIUC your change to the detection op will make this see a tensor of size 1 value 0 here instead of 0 so the asserts below are safe, but itll return a garbage prediction box right?

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (mergennachin) [reply torchvision/ops/boxes.py:121]: It works for empty, let me add a unit test

File: torchvision/models/detection/_utils.py (hunk 2 above)
Comment (mergennachin) [reply torchvision/models/detection/_utils.py]: Yes, good point

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (mergennachin) [reply torchvision/ops/boxes.py:121]: Yes, the downstream handles it correctly. Updated the comment to make this clearer. Also added test_export_zero_detections_structure

File: torchvision/models/detection/_utils.py (hunk 2 above)
Comment (laithsakka) [reply torchvision/models/detection/_utils.py]: you can also have the assertions only on the unbacked path
those will translate  to runtime assertions if not resolved at compile time

something like
  if guard_or_false(box_sum == 0):
     ..
  else:
      torch._check(box_sum > 0, "box_sum assumed to be >0 due to unbacked semantics")
